### PR TITLE
std.net additions

### PR DESCRIPTION
**Pull request by andrewrk**
So far this is DNS resolution stuff.

### CODE REVIEW

File: lib/std/os/bits/linux.zig [hunk 1]
```diff
@@ -1384,3 +1384,85 @@ pub const Statx = extern struct {
...
+pub const EAI_INPROGRESS = -100;
+pub const EAI_CANCELED = -101;
+pub const EAI_NOTCANCELED = -102;
+pub const EAI_ALLDONE = -103;
+pub const EAI_INTR = -104;
+pub const EAI_IDN_ENCODE = -105;
+
+pub const IPPORT_RESERVED = 1024;
+
+pub const IPPROTO_IP = 0;
```
Comment (daurnimator) [inline lib/std/os/bits/linux.zig]: We already have these as the misnamed `PROTO_` constants see https://github.com/ziglang/zig/blob/634f56ee301a8308b2e0b94606cdb5e8151813a6/lib/std/os/bits/linux.zig#L230

File: lib/std/os/bits/linux.zig [hunk 2]
```diff
@@ -1384,3 +1384,85 @@ pub const Statx = extern struct {
...
+pub const EAI_AGAIN = -3;
+pub const EAI_FAIL = -4;
+pub const EAI_FAMILY = -6;
+pub const EAI_SOCKTYPE = -7;
+pub const EAI_SERVICE = -8;
+pub const EAI_MEMORY = -10;
+pub const EAI_SYSTEM = -11;
+pub const EAI_OVERFLOW = -12;
+
+pub const EAI_NODATA = -5;
```
Comment (daurnimator) [inline lib/std/os/bits/linux.zig]: These (and `AI_` and `NI_`) are libc constants and don't belong under os/bits

File: lib/std/special/start.zig [hunk 3]
```diff
@@ -222,6 +222,3 @@ inline fn callMain() u8 {
         else => @compileError(bad_main_ret),
     }
 }
-
-const main_thread_tls_align = 32;
```
Comment (daurnimator) [inline lib/std/special/start.zig]: This change should probably go straight to master and not in this PR?

Comment (andrewrk) [reply lib/std/special/start.zig]: Yeah, sure, I'll cherry pick it out and rebase

**Comment by andrewrk** [comment]
Some notes on the progress, I refreshed the "listen on a port, send bytes, receive bytes" test.

#### example std.net.TcpServer API usage

(this example only works in evented I/O mode):

https://github.com/ziglang/zig/blob/b3c80418a8f8c235e798feb7ee585bc1dcacdf0f/lib/std/net/test.zig#L45-L71

#### std.net.TcpServer implementation

(impl doesn't care if evented or blocking I/O, and notice there are no async or await keywords anywhere): 

https://github.com/ziglang/zig/blob/16397241f6d85956f460f1b1049cf9ce0312bca4/lib/std/net.zig#L1255-L1348

#### strace

```strace
execve("./test", ["./test"], 0x7ffe7d999e00 /* 127 vars */) = 0
rt_sigaction(SIGSEGV, {sa_handler=0x25ba30, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_RESETHAND|SA_SIGINFO, sa_restorer=0x2075b0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6f93073000
eventfd2(1, EFD_CLOEXEC|EFD_NONBLOCK)   = 3
epoll_create1(EPOLL_CLOEXEC)            = 4
eventfd2(0, EFD_CLOEXEC|EFD_NONBLOCK)   = 5
epoll_ctl(4, EPOLL_CTL_ADD, 5, {EPOLLIN, {u32=2531936, u64=2531936}}) = 0
mmap(NULL, 16785408, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6f92071000
mprotect(0x7f6f92071000, 16785408, PROT_READ|PROT_WRITE) = 0
clone(child_stack=0x7f6f93071ff8, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID|0x400000, parent_tid=[28275], child_tidptr=0x7f6f93072000) = 28275
socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_TCP) = 6
bind(6, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
listen(6, 128)                          = 0
getsockname(6, {sa_family=AF_INET, sin_port=htons(53527), sin_addr=inet_addr("127.0.0.1")}, [16]) = 0
accept4(6, 0x7ffe607c7ad0, [112], SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
epoll_ctl(4, EPOLL_CTL_ADD, 6, {EPOLLIN|EPOLLONESHOT|EPOLLET, {u32=1618771536, u64=140730517192272}}) = 0
socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_TCP) = 7
connect(7, {sa_family=AF_INET, sin_port=htons(53527), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
epoll_ctl(4, EPOLL_CTL_ADD, 7, {EPOLLIN|EPOLLOUT|EPOLLONESHOT|EPOLLET, {u32=1618773776, u64=140730517194512}}) = 0
epoll_pwait(4, [{EPOLLIN, {u32=1618771536, u64=140730517192272}}], 1, -1, NULL, 128) = 1
epoll_ctl(4, EPOLL_CTL_DEL, 6, NULL)    = 0
accept4(6, {sa_family=AF_INET, sin_port=htons(52202), sin_addr=inet_addr("127.0.0.1")}, [112->16], SOCK_CLOEXEC|SOCK_NONBLOCK) = 8
write(8, "hello from server\n", 18)     = 18
epoll_pwait(4, [{EPOLLIN|EPOLLOUT, {u32=1618773776, u64=140730517194512}}], 1, -1, NULL, 128) = 1
epoll_ctl(4, EPOLL_CTL_DEL, 7, NULL)    = 0
futex(0x26a1d8, FUTEX_WAKE, 1)          = 1
write(5, "\1\1\1\1\1\1\1\1", 8)         = 8
getsockopt(7, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
read(7, "hello from server\n", 100)     = 18
close(7)                                = 0
close(6)                                = 0
epoll_pwait(4, [{EPOLLIN, {u32=2531936, u64=2531936}}], 1, -1, NULL, 128) = 1
munmap(0x7f6f92071000, 16785408)        = 0
close(5)                                = 0
close(3)                                = 0
close(4)                                = 0
munmap(0x7f6f93073000, 32)              = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

#### next steps

I'm planning to get as far as ability to download a file to disk over a http 1.0 connection, before trying to start merging the branch. Other POSIX's I expect to be relatively easy, however I know next to nothing about Windows networking and would certainly appreciate help with that. I'll probably be looking at libuv for inspiration.

File: lib/std/net.zig [hunk 4]
```diff
@@ -235,13 +294,1055 @@ pub fn connectUnixSocket(path: []const u8) !std.fs.File {
...
+        SystemResources,
+
+        ProtocolFailure,
+
+        /// Firewall rules forbid connection.
+        BlockedByFirewall,
+    } || os.UnexpectedError;
+
+    /// If this function succeeds, the returned `fs.File` is a caller-managed resource.
+    pub fn accept(self: *TcpServer) AcceptError!fs.File {
```
Comment (euantorano) [inline lib/std/net.zig]: It's handy to know the IP address of the accepted socket for a TCP server. Not sure if you already plan to change this return type to include that information or not.

**Comment by euantorano** [comment]
I noticed some changes to IP address parsing in this PR, and thought I'd add some thoughts in case you plan to make more changes here.

It's worth noting all of the following are valid IPv6 addresses that I don't believe will parse as it stands due to [the check here](https://github.com/ziglang/zig/blob/16397241f6d85956f460f1b1049cf9ce0312bca4/lib/std/net.zig#L224):

- `::1`
- `::` (all segments are 0)
- `2001:db8::` (last 6 segments are 0)
- `::1234:5678` (first 6 segments are 0)
- `2001:db8::1234:5678` (middle 4 segments are 0)

I also don't know if you intend to support IPv4 mapped or IPv4 compatible IPv6 addresses, but it would be nice to do so as they're quite common:

- `::ffff:192.0.2.128` (IPv4 mapped IPv6 address)
- `::192.0.2.128` (IPv4 comaptible - though this format is deprecated)

**Pull request merged**